Fix action database picker state without available databases

kulyk opened the pull request:
Epic #27581

### Description

Fixes how the action editor database picker looks like if a user doesn't have access to any database with actions enabled

### How to verify

1. As an admin user, create a model and add a query action to it
2. Create a non-user on `/admin/people`
3. Turn off data permission to your model's database on `/admin/permissions/data/group`
4. Sign in as that user and open an action in the editor
5. Ensure you can't see an empty dropdown next to the database picker

### Demo

**Before**

<img width="352" alt="before" src="https://user-images.githubusercontent.com/17258145/223780764-1a89d17d-de96-4fd1-9676-5745c679095a.png">

**After**

<img width="309" alt="CleanShot 2023-03-15 at 14 11 25@2x" src="https://user-images.githubusercontent.com/17258145/225335022-e29e6a33-aeda-49d6-a033-008a9f7580ff.png">

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/metabase/metabase/29039)
<!-- Reviewable:end -->

kulyk committed:
Fix empty database picker state

kulyk committed:
Remove redundant `requireWriteback` occurrences

kulyk committed:
Update test

kulyk commented on e2e/test/scenarios/models/model-actions.cy.spec.js:
```diff
@@ -174,12 +174,6 @@ describe(
       cy.findByText("New").click();
       popover().findByText("Action").click();
 
-      cy.findByText("Select a database").click();
-      popover().within(() => {
-        cy.findByText("Sample Database").should("not.exist");
-        cy.findByText("QA Postgres12").click();
-      });
-
```
This is no longer needed because the database picker would automatically select the only available option. Before, the picker would have the sample database in the list of databases, but now it gets filtered out at the parent component because actions are not disabled for it

kulyk committed:
Update test

EmmadUsmani commented on frontend/src/metabase/query_builder/components/NativeQueryEditor/DataSourceSelectors/DataSourceSelectors.jsx:
```diff
@@ -60,7 +60,14 @@ const DataSourceSelectors = ({
+    if (!requireWriteback) {
+      return databases;
+    }
+    return databases.filter(database => checkDatabaseActionsEnabled(database));
```
Nit: code can be a bit simpler here if you just pass the function to filter directly

kulyk replied:
I'm a bit traumatized by `list.map(parseInt)`, so TBH I think it's good enough as is

EmmadUsmani commented on frontend/src/metabase/query_builder/components/NativeQueryEditor/DataSourceSelectors/DataSourceSelectors.jsx:
```diff
@@ -60,7 +60,14 @@ const DataSourceSelectors = ({
-  const databases = query.metadata().databasesList({ savedQuestions: false });
+
+  const databases = useMemo(() => {
+    const databases = query.metadata().databasesList({ savedQuestions: false });
```
Nit: might be better to use a different variable name like `queriedDatabases` to avoid confusion with the variable defined on the previous line.

EmmadUsmani reviewed (approved):
Overall looks good! One thing I noticed in the demo screenshot is that the editor says "This **question** is written in SQL." Should we change that to say "action" instead? If so that can be done in a separate PR

kulyk commented:
> One thing I noticed in the demo screenshot is that the editor says "This question is written in SQL."

Great catch @EmmadUsmani, thank you! Let me see if we can do something about this given the translations are locked

kulyk committed:
Use clearer names

kulyk committed:
Don't show query language message for actions

ranquild commented on frontend/src/metabase/query_builder/components/NativeQueryEditor/DataSourceSelectors/DataSourceSelectors.jsx:
```diff
@@ -177,11 +184,18 @@ const TableSelector = ({ database, readOnly, selectedTable, setTableId }) => (
-  </div>
-);
+const Placeholder = ({ query, editorContext }) => {
+  if (editorContext === "action") {
```
nit: Maybe `This action is written in `${language}.`?

ranquild replied:
If we don't want to add new translations, maybe return just `language` here? And we can add a proper placeholder in master only

ranquild commented on frontend/src/metabase/query_builder/components/NativeQueryEditor/DataSourceSelectors/DataSourceSelectors.jsx:
```diff
@@ -58,12 +60,26 @@ const DataSourceSelectors = ({
+    }
+
+    return allDatabases.filter(database =>
+      checkDatabaseActionsEnabled(database),
```
`database.hasActionsEnabled()`

iethree replied:
so, a weird consequence of this filter behavior for non-admin users is we let them view queries for databases that they don't have access to. This is probably a minor correctness issue, but if you have access to a write-enabled DB, and view a query action for another write-enabled db, the action editor will incorrectly show you the name of the DB you have access to.

Example:

Admin | limited user
--|--
![Screen Shot 2023-03-13 at 10 07 12 AM](https://user-images.githubusercontent.com/30528226/224759694-b5c93dfd-7eab-46fe-a295-508b3ed0a3f0.png) | ![Screen Shot 2023-03-13 at 10 07 18 AM](https://user-images.githubusercontent.com/30528226/224759699-5d9aeefc-99ef-4a9f-9284-cc34358a3d61.png)

ranquild commented on frontend/src/metabase/query_builder/components/NativeQueryEditor/DataSourceSelectors/DataSourceSelectors.jsx:
```diff
@@ -58,12 +60,26 @@ const DataSourceSelectors = ({
   setDatabaseId,
   setTableId,
   requireWriteback = false,
```
Should it be removed?

ranquild commented on frontend/src/metabase/query_builder/components/NativeQueryEditor/DataSourceSelectors/DataSourceSelectors.jsx:
```diff
@@ -58,12 +60,26 @@ const DataSourceSelectors = ({
+      .metadata()
+      .databasesList({ savedQuestions: false });
+
+    if (!requireWriteback) {
```
`editorContext === "action"`

ranquild reviewed (commented):
@kulyk I think if it's hard to add a unit test for this (and it is hard), we should add a e2e test instead

iethree commented on frontend/src/metabase/query_builder/components/NativeQueryEditor/DataSourceSelectors/DataSourceSelectors.jsx:
```diff
@@ -135,21 +149,14 @@ const checkIfThereAreMultipleDatabases = (database, databases) =>
   database == null ||
   (databases.length > 1 && databases.some(db => db.id === database.id));
```
we might need to split this check into 2 checks, and have a case for when the selected db isn't int he list of dbs.

kulyk committed:
Use database `hasActionsEnabled` method

kulyk committed:
Remove `requireWriteback` prop

kulyk committed:
Respect `readOnly` prop in `DataSourceSelectors`

kulyk committed:
Extend e2e test coverage

kulyk committed:
Use `setActionsEnabledForDB` in Cypress

kulyk committed:
Fix data picker behavior

kulyk commented:
@iethree @ranquild @EmmadUsmani, thanks for your reviews! I think I addressed all of your comments and also extended one of the e2e tests to cover these changes better. Could you please take another look at the PR?

iethree reviewed (approved):
![chefs-kiss-french-chef](https://user-images.githubusercontent.com/30528226/225446526-8b75ae46-db43-4401-9151-70a04f11d17d.gif)
